### ISSUE TITLE
Volume transformer computes bad multiplier

### ISSUE THREAD

**Issue by meishuu** [hydrabolt/discord.js#242]
**tl;dr**: `Math.tan(this.volume)` is terrible. Let's try `Math.pow(this.volume, 1.661)` and go from there.

## Intro

Since loudness is logarithmic rather than linear, we need a way to map a volume factor from [0, 1] to an amplitude multiplier such that the resulting scaled value best approximates our desired volume amount.

At minimum, we would like a function f to compute a multiplier such that f(0) = 0 and f(1) = 1.

[Currently](https://github.com/hydrabolt/discord.js/blob/81549e7/src/Voice/VolumeTransformer.js#L21), the multiplier in `VolumeTransformer` is computed as `Math.tan(this.volume)`. While it satisfies f(0) = 0, we have the unfortunate effect that f(1) = tan(1) ≈ 1.5574. In other words, **at the default volume 1, samples are being multiplied by 1.5574**.

Tracing the credits back to the original author, this was noted for the use of tangent as the scaling function:

>`float multiplier = tan(some_level/100.0);`

>If some_level is set to a value between 0 and 148 or so, this will give you a rather linear sounding multiplier. 79 is almost a multiplier of 1 (no amplification). It is far -- really far -- from perfect, but it worked well enough for my needs of implementing a volume slider.

Thus, if we are to keep using tangent, we should be using `Math.tan(this.volume * 0.79)` instead. But this is pretty hacky, so let's not do that.

---

## Analysis

Decibels seem to be easier to work with, so let's figure out a way to map a volume factor to decibels.

I'm not really sure what good values to use here are, but let's assume that half (0.5) volume should mean -10 dB and double (2) volume is +10 dB. (Of course, 1 volume = 0 dB.)

`db = 10 * log2(volume)` satisfies this, so we'll use that.

Now, we can convert decibels to an amplitude ratio using [this formula](https://en.wikipedia.org/wiki/Decibel#Field_quantities_and_root-power_quantities):

`amplitude_ratio = 10 ^ (db / 20)`

Plugging it in:

`amplitude_ratio = 10 ^ ([10 * log2(volume)] / 20)`
`... = 10 ^ [0.5 * log10(volume) / log10(2)]`
`... = volume ^ [0.5 / log10(2)]`
`... ≈ volume ^ 1.660964`

Testing this:

volume | tan(x) | x^1.661
--- | --- | ---
0 | 0 | 0
0.5 | 0.5463 | 0.3162
1 | 1.5574 | 1
2 | -2.1850 | 3.1624

Note that this suggested formula also has the advantage of working well for values above 1, and that the mapping of decibels to amplitude ratios is correct as given by the [Wikipedia table](https://en.wikipedia.org/wiki/Decibel).

---

## Conclusion

Wanna be lazy? I'll suggest `Math.pow(this.volume, 1.661)` in place of tangent.

Wanna do better? Add a way to give a decibel value just like before, and then use `Math.pow(10, db / 20)` to convert to the multiplier. This is probably ideal.

**Comment by aequasi** [comment]
Thank you for this writeup. I implemented the volume stuff before, and really had no idea what I was doing haha.

**Comment by meew0** [comment]
I'm going to close this issue, @meishuu if there's still a problem with the stuff, reopen it
